File: 49-group-anagrams/group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> mp;
        for(auto s : strs){
            string str = s;
            sort(str.begin(), str.end());
            mp[str].push_back(s);
        }
        vector<vector<string>> res;
        for(auto m : mp){
            res.push_back(m.second);
        }
        return res;
    }
};

File: 49-group-anagrams/group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> res = new ArrayList<>();
        HashMap<String, ArrayList<String>> mp = new HashMap<>();
        for(int i=0; i<strs.length; i++){
            int[] temp = new int[26];
            String curr = strs[i];
            for(int j=0; j<curr.length(); j++){
                temp[curr.charAt(j) - 'a']++;
            }
            StringBuilder sb = new StringBuilder("");
            for(int k = 0; k<26; k++){
                int count = temp[k];
                while(count > 0){
                    sb.append('a' + k);
                    count--;
                }
            }

            String key = sb.toString();

            if(!mp.containsKey(key)) {
                mp.put(key, new ArrayList<>());
            }

            mp.get(key).add(strs[i]);
        }
        for(Map.Entry<String, ArrayList<String>> e:mp.entrySet()){
            res.add(e.getValue());
        }
        return res;
    }
}